main/src/library/MutDisjointSet.flix

```diff
@@ -0,0 +1,149 @@
+/*
+ *  Copyright 2023 Jakob Schneider Villumsen
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+pub enum MutDisjointSet[t: Type, r: Eff]({
+    rc = Region[r],
+    forest = MutMap[t, MutDisjointSet.Node[t, r], r]
+})
+
+mod MutDisjointSet {
+
+    pub enum Node[t: Type, r: Eff]({
+        rc = Region[r],
+        value = t,
+        parent = Ref[Option[Node[t, r]], r],
+        rank = Ref[Int32, r]
+    })
+
+    ///
+    /// Returns a new empty `MutDisjointSet` in the region `rc`.
+    ///
+    pub def new(rc: Region[r]): MutDisjointSet[t, r] \ r =
+        MutDisjointSet({ rc = rc, forest = MutMap.new(rc) })
+
+    ///
+    /// Updates `s` with a new disjoint set containing `x`
+    /// if `x` is not already in `s`.
+    ///
+    pub def makeSet(x: t, s: MutDisjointSet[t, r]): Unit \ r with Order[t] = {
+        let MutDisjointSet({ rc, forest }) = s;
+        if (not contains(x, s)) {
+            MutMap.put!(x, singleton(rc, x), forest)
+        } else {
+            ()
+        }
+    }
+
+    ///
+    /// Returns `true` iff `s` contains `x`.
+    ///
+    pub def contains(x: t, s: MutDisjointSet[t, r]): Bool \ r with Order[t] = {
+        let MutDisjointSet({ forest | _ }) = s;
+        match MutMap.get(x, forest) {
+            case Some(_) => true
+            case None    => false
+        }
+    }
+
+    ///
+    /// Returns a new `Node` from `x` in `rc`.
```

Comment:
Returns ... a new disjoint forest with a single element ?

Reply:
No, because that's not what it does :)
We can rename it, though, to `node` and make `singleton` an alias for `makeSet`.

Reply:
I overlooked that this was private. Perhaps we can make it a local function where it is used?

Reply:
Because collections commonly have a singleton function-- but as you say, this is not that.

Reply:
Okay, but for usability, should we then still add `singleton` as suggested above?

Reply:
Mjah. There are many functions we could -- in a subsequent PR.